src/shims/ffi_support.rs

```
@@ -0,0 +1,289 @@
+        // On linux `libloading` is based on `dlsym`: https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#202
+        // and `dlsym`(https://linux.die.net/man/3/dlsym) looks through the dependency tree of the
+        // library if it can't find the symbol in the library itself.
+        // So, in order to check if the function was actually found in the specified
+        // `machine.external_so_lib` we need to check its `dli_fname` and compare it to
+        // the specified SO file path.
+        // This code is a reimplementation of the mechanism for getting `dli_fname` in `libloading`,
+        // from: https://docs.rs/libloading/0.7.3/src/libloading/os/unix/mod.rs.html#411
+        // using the `libc` crate where this interface is public.
+        // No `libc::dladdr` on windows.
+        #[cfg(unix)]
+        let mut info = std::mem::MaybeUninit::<libc::Dl_info>::uninit();
+        #[cfg(unix)]
+        unsafe {
+            if libc::dladdr(*func.deref() as *const _, info.as_mut_ptr()) != 0 {
+                if std::ffi::CStr::from_ptr(info.assume_init().dli_fname).to_str().unwrap()
+                    != _lib_path.to_str().unwrap()
+                {
+                    return None;
+                }
+            }
+        }
+        // Return a pointer to the function.
+        Some(CodePtr(*func.deref() as *mut _))
+    }
+
+    /// Call specified external C function, with supplied arguments.
+    /// Need to convert all the arguments from their hir representations to
+    /// a form compatible with C (through `libffi` call).
+    /// Then, convert return from the C call into a corresponding form that
+    /// can be stored in Miri internal memory.
+    fn call_external_c_fct(
+        &mut self,
+        link_name: Symbol,
+        dest: &PlaceTy<'tcx, Provenance>,
+        args: &[OpTy<'tcx, Provenance>],
+    ) -> InterpResult<'tcx, bool> {
+        // Get the pointer to the function in the shared object file if it exists.
```

It'd probably be a good idea to start with a check here comparing `env!("TARGET")` (which is the host target, i.e., what Miri was built for) and... uh somewhere in `tcx.sess.target` we should be able to get a `TargetTriple` that we can call `str` on to compare this with.

---

Ah, `tcx.sess.opts.target_triple`  seems to be it.

---

Ah, makes sense -- i added this as an extra check in `emulate_foreign_item_by_name`, before even trying to call the external C function. If they don't match it just continues to try the shims. D'you think it should error instead?

---

Ah, I see. I think that check should live in this file.

But yeah it'd also make sense to fail loudly rather than silently, since this is probably not the intended behavior.